pkg/image/pull_policy.go

```diff
@@ -1,41 +1,296 @@
 package image
 
 import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
 	"github.com/pkg/errors"
+
+	"github.com/buildpacks/pack/internal/config"
+	"github.com/buildpacks/pack/pkg/logging"
 )
 
 // PullPolicy defines a policy for how to manage images
 type PullPolicy int
 
+var interval string
+
+type ImagePullPolicyManager struct {
+	Logger logging.Logger
+}
+
+var (
+	hourly        = "1h"
+	daily         = "1d"
+	weekly        = "7d"
+	intervalRegex = regexp.MustCompile(`^(\d+d)?(\d+h)?(\d+m)?$`)
+)
+
 const (
-	// PullAlways images, even if they are present
+	// Always pull images, even if they are present
 	PullAlways PullPolicy = iota
-	// PullNever images, even if they are not present
+	// Pull images hourly
+	PullHourly
+	// Pull images daily
+	PullDaily
+	// Pull images weekly
+	PullWeekly
+	// Never pull images, even if they are not present
 	PullNever
 	// PullIfNotPresent pulls images if they aren't present
 	PullIfNotPresent
+	// PullWithInterval pulls images with specified intervals
+	PullWithInterval
 )
 
-var nameMap = map[string]PullPolicy{"always": PullAlways, "never": PullNever, "if-not-present": PullIfNotPresent, "": PullAlways}
+type Interval struct {
+	PullingInterval string `json:"pulling_interval"`
+	PruningInterval string `json:"pruning_interval"`
+	LastPrune       string `json:"last_prune"`
+}
+
+type ImageData struct {
+	ImageIDtoTIME map[string]string
+}
+
+type ImageJSON struct {
+	Interval *Interval  `json:"interval"`
+	Image    *ImageData `json:"image"`
+}
+
+func DefaultImageJSONPath() (string, error) {
+	home, err := config.PackHome()
+	if err != nil {
+		return "", errors.Wrap(err, "getting pack home")
+	}
+	return filepath.Join(home, "image.json"), nil
+}
+
+var nameMap = map[string]PullPolicy{"always": PullAlways, "hourly": PullHourly, "daily": PullDaily, "weekly": PullWeekly, "never": PullNever, "if-not-present": PullIfNotPresent, "": PullAlways}
+
+// ParsePullPolicy from string with support for interval formats
+func ParsePullPolicy(policy string, logger logging.Logger) (PullPolicy, error) {
+	pullPolicyManager := NewPullPolicyManager(logger)
```

Comment:
@jjbustamante I'm not sure about spawning the manager here. should we create the manager in the caller function of `ParsePullPolicy` and then pass it to the function?

Reply:
Yeah, let add some suggestions about this. 

First of all, a simplify class diagram of your current implementation looks like this:

```mermaid
classDiagram
    Client o-- ImageFetcher
    Fetcher  ..|>  ImageFetcher
    Fetcher o-- ImagePullChecker 
    ImagePullPolicyManager ..|> ImagePullChecker

    Client : +bool experimental
    Client : +map[string]string registryMirrors
    Client : +string version
    Client : +authn.Keychain

    Client --> ImagePullPolicyManager: Remove it!
    ImagePullChecker ..> Fetcher : Remove it!
  
    class  ImageFetcher {
          <<interface>>
    }

    class Fetcher {
    }

    class ImagePullChecker {
        <<interface>>
        CheckImagePullInterval()
	Read()
	PruneOldImages(f *Fetcher) error
	UpdateImagePullRecord()
        Write()
    }

    class ImagePullPolicyManager {
    }
```

The first thing I noticed that doesn't look very health is the circular dependency between `ImagePullChecker` and `Fetcher` in the method `PruneOldImages(f *Fetcher) error`, we need to remove that, I noticed what you are trying to do is to load a local image from the daemon to check if the image exists, maybe you can use the `imgutil.Local` image to do that, but we need to remove that circular dependency.

The other thing I will suggest is:
- You are creating an instance of `NewPullPolicyManager(logger)` in 4 places right now, I think we can:
   -  hold a reference into the Client to an `image.ImagePullChecker` and add an option to set it up
   ```go
type Client struct {
   imagePullChecker    image.ImagePullChecker
}

func WithImagePullChecker(i image.ImagePullChecker) Option {
       return func(c *Client) {
               c.imagePullChecker = i
       }
}
```
- After that, you can move the creation of the PullPolicyManager to the `cmd/cmd.go` just right before creating the client. Update the method to create the client to use the `Option` we declared before
```go
imagePullChecker := image.NewPullPolicyManager(logger)
packClient, err := initClient(logger, cfg, imagePullChecker)

func initClient(logger logging.Logger, cfg config.Config, imagePullChecker image.ImagePullChecker) (*client.Client, error) {
 return client.NewClient(client.WithLogger(logger), client.WithExperimental(cfg.Experimental), ... client.WithImagePullChecker(imagePullChecker))
}
```
- I will move the `ParsePullPolicy` method to be part of the `ImagePullChecker` interface
```go
func (i *ImagePullPolicyManager) ParsePullPolicy(policy string) (PullPolicy, error) {
  // Do what we have today
}
```
- Because of the previous change, all the commands are broken now, I will update the commands that need to call  `ParsePullPolicy` and passthrough the `ImagePullChecker`, for example in `cmd/cmd.go`
```go
 // remember we have the imagePullChecker at this point, because cmd is creating it.
 rootCmd.AddCommand(commands.Build(logger, cfg, packClient, imagePullChecker))
```
- I will repeat the same thing for all the other commands that are broken